### ISSUE TITLE
problem during mpich2 make with slurm and pmi2

### ISSUE THREAD

**Issue by praveenvatsh** [pmodels/mpich#6060]
src/util/mpir_pmi.c:842:53: error: unknown type name 'PMI_keyval_t'
  842 | static int mpi_to_pmi_keyvals(MPIR_Info * info_ptr, PMI_keyval_t ** kv_ptr, int *nkeys_ptr);
      |                                                     ^~~~~~~~~~~~
src/util/mpir_pmi.c:843:30: error: unknown type name 'PMI_keyval_t'
  843 | static void free_pmi_keyvals(PMI_keyval_t ** kv, int size, int *counts);
      |                              ^~~~~~~~~~~~
src/util/mpir_pmi.c:1219:53: error: unknown type name 'PMI_keyval_t'
 1219 | static int mpi_to_pmi_keyvals(MPIR_Info * info_ptr, PMI_keyval_t ** kv_ptr, int *nkeys_ptr)
      |                                                     ^~~~~~~~~~~~
src/util/mpir_pmi.c:1256:30: error: unknown type name 'PMI_keyval_t'
 1256 | static void free_pmi_keyvals(PMI_keyval_t ** kv, int size, int *counts)
      |                              ^~~~~~~~~~~~
make[2]: *** [src/util/lib_libmpi_la-mpir_pmi.lo] Error 1
make[2]: Leaving directory `/home/psingh.iitr/library/mpich2/mpich-4.0.1'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/psingh.iitr/library/mpich2/mpich-4.0.1'
make: *** [all] Error 2

**Comment by hzhou** [comment]
This has been fixed in the current main branch and in 4.1a1.

mpich-4.0.1 should work with Slurm with the default PMI-1. @praveenvatsh What is your configure line?

**Comment by praveenvatsh** [comment]
@hzhou 

./configure --prefix=/home/psingh.iitr/library/mpich2/mpich2.install --with-slurm=/usr/include/slurm --with-pmi=pmi2 --with-pm=no --with-pmi-libdir=/usr/lib64/slurm CPPFLAGS=-I/usr/include/slurm LDFLAGS=-L/usr/lib64/slurm

Got the same error with the ver 4.0.2, also.

**Comment by praveenvatsh** [comment]
@hzhou 

[psingh.iitr@login06 mpich-4.1a1]$ srun --mpi=list
srun: MPI types are...
srun: cray_shasta
srun: none
srun: pmi2

I need pmi2 because pmi1 is not available. 

Got error with ver 4.1a1 also,

make[2]: Entering directory `/home/psingh.iitr/library/mpich2/mpich-4.1a1'
  CC       src/util/lib_libmpi_la-mpir_pmi.lo
src/util/mpir_pmi.c: In function 'MPIR_pmi_set_threaded':
src/util/mpir_pmi.c:252:5: warning: implicit declaration of function 'PMI2_Set_threaded' [-Wimplicit-function-declaration]
  252 |     PMI2_Set_threaded(is_threaded);
      |     ^~~~~~~~~~~~~~~~~
src/util/mpir_pmi.c: At top level:
src/util/mpir_pmi.c:86:19: error: unknown type name 'PMI2_keyval_t'; did you mean 'PMI_keyval_t'?
   86 | #define INFO_TYPE PMI2_keyval_t
      |                   ^~~~~~~~~~~~~
src/util/mpir_pmi.c:918:53: note: in expansion of macro 'INFO_TYPE'
  918 | static int mpi_to_pmi_keyvals(MPIR_Info * info_ptr, INFO_TYPE ** kv_ptr, int *nkeys_ptr);
      |                                                     ^~~~~~~~~
src/util/mpir_pmi.c:86:19: error: unknown type name 'PMI2_keyval_t'; did you mean 'PMI_keyval_t'?
   86 | #define INFO_TYPE PMI2_keyval_t
      |                   ^~~~~~~~~~~~~
src/util/mpir_pmi.c:919:30: note: in expansion of macro 'INFO_TYPE'
  919 | static void free_pmi_keyvals(INFO_TYPE ** kv, int size, int *counts);
      |                              ^~~~~~~~~
src/util/mpir_pmi.c: In function 'MPIR_pmi_spawn_multiple':
src/util/mpir_pmi.c:86:19: error: unknown type name 'PMI2_keyval_t'; did you mean 'PMI_keyval_t'?
   86 | #define INFO_TYPE PMI2_keyval_t
      |                   ^~~~~~~~~~~~~
src/util/mpir_pmi.c:934:5: note: in expansion of macro 'INFO_TYPE'
  934 |     INFO_TYPE **info_keyval_vectors = NULL;
      |     ^~~~~~~~~
src/util/mpir_pmi.c:86:19: error: unknown type name 'PMI2_keyval_t'; did you mean 'PMI_keyval_t'?
   86 | #define INFO_TYPE PMI2_keyval_t
      |                   ^~~~~~~~~~~~~
src/util/mpir_pmi.c:939:5: note: in expansion of macro 'INFO_TYPE'
  939 |     INFO_TYPE *preput_vector = NULL;
      |     ^~~~~~~~~
src/util/mpir_pmi.c:86:19: error: 'PMI2_keyval_t' undeclared (first use in this function); did you mean 'PMI_keyval_t'?
   86 | #define INFO_TYPE PMI2_keyval_t
      |                   ^~~~~~~~~~~~~
src/util/mpir_pmi.c:945:28: note: in expansion of macro 'INFO_TYPE'
  945 |     info_keyval_vectors = (INFO_TYPE **) MPL_malloc(count * sizeof(INFO_TYPE *), MPL_MEM_BUFFER);
      |                            ^~~~~~~~~
src/util/mpir_pmi.c:86:19: note: each undeclared identifier is reported only once for each function it appears in
   86 | #define INFO_TYPE PMI2_keyval_t
      |                   ^~~~~~~~~~~~~
src/util/mpir_pmi.c:945:28: note: in expansion of macro 'INFO_TYPE'
  945 |     info_keyval_vectors = (INFO_TYPE **) MPL_malloc(count * sizeof(INFO_TYPE *), MPL_MEM_BUFFER);
      |                            ^~~~~~~~~
src/util/mpir_pmi.c:945:40: error: expected expression before ')' token
  945 |     info_keyval_vectors = (INFO_TYPE **) MPL_malloc(count * sizeof(INFO_TYPE *), MPL_MEM_BUFFER);
      |                                        ^
src/util/mpir_pmi.c:955:25: warning: implicit declaration of function 'mpi_to_pmi_keyvals' [-Wimplicit-function-declaration]
  955 |             mpi_errno = mpi_to_pmi_keyvals(info_ptrs[i], &info_keyval_vectors[i],
      |                         ^~~~~~~~~~~~~~~~~~
src/util/mpir_pmi.c:87:31: error: request for member 'key' in something not a structure or union
   87 | #define INFO_TYPE_KEY(kv) (kv).key
      |                               ^
src/util/mpir_pmi.c:976:13: note: in expansion of macro 'INFO_TYPE_KEY'
  976 |             INFO_TYPE_KEY(preput_vector[i]) = preput_keyvals[i].key;
      |             ^~~~~~~~~~~~~
src/util/mpir_pmi.c:88:31: error: request for member 'val' in something not a structure or union
   88 | #define INFO_TYPE_VAL(kv) (kv).val
      |                               ^
src/util/mpir_pmi.c:977:13: note: in expansion of macro 'INFO_TYPE_VAL'
  977 |             INFO_TYPE_VAL(preput_vector[i]) = preput_keyvals[i].val;
      |             ^~~~~~~~~~~~~
src/util/mpir_pmi.c:1016:38: error: expected ')' before 'PMI2_keyval_t'
 1016 |                                (const INFO_TYPE **) info_keyval_vectors,
      |                                ~     ^
      |                                      )
src/util/mpir_pmi.c:1016:32: warning: type defaults to 'int' in type name [-Wimplicit-int]
 1016 |                                (const INFO_TYPE **) info_keyval_vectors,
      |                                ^
src/util/mpir_pmi.c:1016:32: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
src/util/mpir_pmi.c:1016:32: warning: passing argument 7 of 'PMI2_Job_Spawn' makes pointer from integer without a cast [-Wint-conversion]
 1016 |                                (const INFO_TYPE **) info_keyval_vectors,
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                |
      |                                int
In file included from ./src/include/mpir_pmi.h:28,
                 from src/util/mpir_pmi.c:6:
/usr/include/slurm/pmi2.h:339:44: note: expected 'const struct MPID_Info **' but argument is of type 'int'
  339 |                    const struct MPID_Info *info_keyval_vectors[],
      |                    ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
src/util/mpir_pmi.c:1017:51: warning: passing argument 9 of 'PMI2_Job_Spawn' from incompatible pointer type [-Wincompatible-pointer-types]
 1017 |                                num_preput_keyval, preput_vector, NULL, 0, pmi_errcodes);
      |                                                   ^~~~~~~~~~~~~
      |                                                   |
      |                                                   int *
In file included from ./src/include/mpir_pmi.h:28,
                 from src/util/mpir_pmi.c:6:
/usr/include/slurm/pmi2.h:341:44: note: expected 'const struct MPID_Info **' but argument is of type 'int *'
  341 |                    const struct MPID_Info *preput_keyval_vector[],
      |                    ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
src/util/mpir_pmi.c:1029:9: warning: implicit declaration of function 'free_pmi_keyvals' [-Wimplicit-function-declaration]
 1029 |         free_pmi_keyvals(info_keyval_vectors, count, info_keyval_sizes);
      |         ^~~~~~~~~~~~~~~~
src/util/mpir_pmi.c: At top level:
src/util/mpir_pmi.c:86:19: error: unknown type name 'PMI2_keyval_t'; did you mean 'PMI_keyval_t'?
   86 | #define INFO_TYPE PMI2_keyval_t
      |                   ^~~~~~~~~~~~~
src/util/mpir_pmi.c:1357:53: note: in expansion of macro 'INFO_TYPE'
 1357 | static int mpi_to_pmi_keyvals(MPIR_Info * info_ptr, INFO_TYPE ** kv_ptr, int *nkeys_ptr)
      |                                                     ^~~~~~~~~
src/util/mpir_pmi.c:86:19: error: unknown type name 'PMI2_keyval_t'; did you mean 'PMI_keyval_t'?
   86 | #define INFO_TYPE PMI2_keyval_t
      |                   ^~~~~~~~~~~~~
src/util/mpir_pmi.c:1398:30: note: in expansion of macro 'INFO_TYPE'
 1398 | static void free_pmi_keyvals(INFO_TYPE ** kv, int size, int *counts)
      |                              ^~~~~~~~~
make[2]: *** [src/util/lib_libmpi_la-mpir_pmi.lo] Error 1
make[2]: Leaving directory `/home/psingh.iitr/library/mpich2/mpich-4.1a1'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/psingh.iitr/library/mpich2/mpich-4.1a1'
make: *** [all] Error 2

used the same configure line posted above.

**Comment by hzhou** [comment]
> @hzhou
> 
> ./configure --prefix=/home/psingh.iitr/library/mpich2/mpich2.install --with-slurm=/usr/include/slurm --with-pmi=pmi2 --with-pm=no --with-pmi-libdir=/usr/lib64/slurm CPPFLAGS=-I/usr/include/slurm LDFLAGS=-L/usr/lib64/slurm
> 
> Got the same error with the ver 4.0.2, also.

Could you try 
```
/configure --prefix=[install-path] --with-pmi=slurm --with-pm=no --with-slurm=/usr/include/slurm --with-pmi-libdir=/usr/lib64/slurm 
```

> [psingh.iitr@login06 mpich-4.1a1]$ srun --mpi=list
> srun: MPI types are...
> srun: cray_shasta
> srun: none
> srun: pmi2
>
> I need pmi2 because pmi1 is not available.

Slurm supports both pmi-1 and pmi-2. They use `pmi2` to refer to both.

**Comment by praveenvatsh** [comment]
> > @hzhou
> > ./configure --prefix=/home/psingh.iitr/library/mpich2/mpich2.install --with-slurm=/usr/include/slurm --with-pmi=pmi2 --with-pm=no --with-pmi-libdir=/usr/lib64/slurm CPPFLAGS=-I/usr/include/slurm LDFLAGS=-L/usr/lib64/slurm
> > Got the same error with the ver 4.0.2, also.
> 
> Could you try
> 
> ```
> /configure --prefix=[install-path] --with-pmi=slurm --with-pm=no --with-slurm=/usr/include/slurm --with-pmi-libdir=/usr/lib64/slurm 
> ```
> 
> > [psingh.iitr@login06 mpich-4.1a1]$ srun --mpi=list
> > srun: MPI types are...
> > srun: cray_shasta
> > srun: none
> > srun: pmi2
> > I need pmi2 because pmi1 is not available.
> 
> Slurm supports both pmi-1 and pmi-2. They use `pmi2` to refer to both.

make[3]: Leaving directory `/home/psingh.iitr/library/mpich2/mpich-4.0.2/src/mpi/romio'
make[2]: Leaving directory `/home/psingh.iitr/library/mpich2/mpich-4.0.2/src/mpi/romio'
Making all in .
make[2]: Entering directory `/home/psingh.iitr/library/mpich2/mpich-4.0.2'
  CC       src/binding/c/attr/lib_libmpi_la-attr_delete.lo
In file included from ./src/include/mpiimpl.h:82,
                 from src/binding/c/attr/attr_delete.c:8:
./src/include/mpir_pmi.h:16:10: fatal error: pmi.h: No such file or directory
   16 | #include <pmi.h>
      |          ^~~~~~~
compilation terminated.
make[2]: *** [src/binding/c/attr/lib_libmpi_la-attr_delete.lo] Error 1
make[2]: Leaving directory `/home/psingh.iitr/library/mpich2/mpich-4.0.2'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/psingh.iitr/library/mpich2/mpich-4.0.2'
make: *** [all] Error 2

**Comment by hzhou** [comment]
The slurm `pmi.h` is usually packaged separately. For Ubuntu, try `apt install libpmi0-dev`.

With mpich-4.1a1, you can also try `--with-pmi=pmi2 --with-pmilib=slurm` to use pmi2 with Slurm. While it works, last time I checked, slurm will spew warning messages due to some internal glitches.

**Comment by praveenvatsh** [comment]
@hzhou 

Thank you
Problem is resolved

**Comment by bilias** [comment]
Is this solved with mpich-4.1a1 or with mpich-4.0.2 as well?

--with-pmi-libdir is not listed in configure of 4.0.2, neither --with-pmilib (I guess these are new options).
--with-pmi=slurm loads pmi.h and not pmi2.h, at least this is what I see in src/pmi/slurm/subconfigure.m4

  CC       src/util/lib_libmpi_la-mpir_pmi.lo
src/util/mpir_pmi.c:842:53: error: unknown type name ‘PMI_keyval_t’; did you mean ‘MPII_Keyval’?                                                                                                                      
 static int mpi_to_pmi_keyvals(MPIR_Info * info_ptr, PMI_keyval_t ** kv_ptr, int *nkeys_ptr);                                                                                                     
                                                 ^~~~~~~~~~~~                                                           
                                                 MPII_Keyval                                 

Reading /usr/include/slurm/pmi2.h (from slurm-devel-22.05.0-1.el8.x86_64)
I see that struct PMI_keyval_t is commented out.

/* This is here to allow spawn multiple functions to compile.  This
   needs to be removed once those functions are fixed for pmi2 */
/*
typedef struct PMI_keyval_t
{
    char * key;
    char * val;
} PMI_keyval_t;
*/

I've changed src/util/mpir_pmi.c:
840: #if defined(USE_PMI1_API) || defined(USE_PMI2_API)
to
#if defined(USE_PMI1_API)
same in line 1266

and build finished with:

export CFLAGS="${CFLAGS} -I/usr/include/slurm"
export CPPFLAGS="${CPPFLAGS} -I/usr/include/slurm"
export LIBS="-lpmi2"
and
--with-slurm=/usr/include/slurm \
--with-pmi=pmi2 \
--with-pm=none \

**Comment by hzhou** [comment]
@bilias This is only fixed in the main branch and mpich-4.1a1. For mpich-4.0.2 and older releases, your suggested patch should work. However, I would recommend simply configure mpich with `--with-pmi=slurm`, which, I believe, will use use PMI1. What Slurm calls "PMI2" in fact includes both PMI1 and PMI2, and to my knowledge, PMI1 works more stable than PMI2.
